docs/dev/table/sql.md

```diff
@@ -813,6 +813,222 @@ LIMIT 3
+
+The optimization way is omitting rownum field in the outer SELECT clause of the Top-N query. This is reasonable because the number of the top N records is usually not large, thus the consumers can sort the records themselves quickly. Without rownum field, in the example above, only the changed record (`product-1001`) needs to be sent to downstream, which can reduce much IO to the result table.
+
+The following example shows how to optimize the above Top-N example in this way:
+
+<div class="codetabs" markdown="1">
+<div data-lang="java" markdown="1">
+{% highlight java %}
+StreamExecutionEnvironment env = StreamExecutionEnvironment.getExecutionEnvironment();
+StreamTableEnvironment tableEnv = TableEnvironment.getTableEnvironment(env);
+
+// ingest a DataStream from an external source
+DataStream<Tuple3<String, String, String, Long>> ds = env.addSource(...);
+// register the DataStream as table "ShopSales"
+tableEnv.registerDataStream("ShopSales", ds, "product_id, category, product_name, sales");
+
+// select top-5 products per category which have the maximum sales.
+Table result1 = tableEnv.sqlQuery(
+  "SELECT product_id, category, product_name, sales " + // omit row_num field in the output
+  "FROM (" +
+  "   SELECT *," +
+  "       ROW_NUMBER() OVER (PARTITION BY category ORDER BY sales DESC) as row_num" +
+  "   FROM ShopSales)" +
+  "WHERE row_num <= 5");
+{% endhighlight %}
+</div>
+
+<div data-lang="scala" markdown="1">
+{% highlight scala %}
+val env = StreamExecutionEnvironment.getExecutionEnvironment
+val tableEnv = TableEnvironment.getTableEnvironment(env)
+
+// read a DataStream from an external source
+val ds: DataStream[(String, String, String, Long)] = env.addSource(...)
+// register the DataStream under the name "ShopSales"
+tableEnv.registerDataStream("ShopSales", ds, 'product_id, 'category, 'product_name, 'sales)
+
+
+// select top-5 products per category which have the maximum sales.
+val result1 = tableEnv.sqlQuery(
+    """
+      |SELECT product_id, category, product_name, sales  -- omit row_num field in the output
+      |FROM (
+      |   SELECT *,
+      |       ROW_NUMBER() OVER (PARTITION BY category ORDER BY sales DESC) as row_num
+      |   FROM ShopSales)
+      |WHERE row_num <= 5
+    """.stripMargin)
+{% endhighlight %}
+</div>
+</div>
+
+<span class="label label-danger">Attention</span> In order to output the above query to an external storage and have a correct result, the external storage must have the same primary key with the Top-N query. In the above example query, if the `product_id` is the primary key of the query, then the external table should also has `product_id` as the primary key.
```

Comment:
`if the top N records need to be stored in external storage, the result table must have the same primary key with the Top-N query. By default, the primary key of Top-N query is the combination of partition columns + rownum column.` above.
Confuse here, so what is the real pks of top-n query?

Reply:
I mean the pk of topn query is (partition columns + rownum column), for example, if partition column is `a`, `b`,  so the pk of topn is the combination of `a`, `b`, `rownum`

Reply:
Oh, I see. When the user omit the rownumber, it can't be output directly to the external storage with pk.

Reply:
It can be explained that the PK changes caused by optimization and non-optimization are difficult to understand.

Reply:
> It can be explained that the PK changes caused by optimization and non-optimization are difficult to understand.

+1, the top-N implements with optimization and non-optimization are difficult to understand according to the description.